File: go/hard.go
package main

import (
	"fmt"
	"net/http"
	"sync"
)

var balances = map[string]int{
	"alice": 100,
	"bob":   50,
}

var mu sync.Mutex


func transfer(from, to string, amount int) {
	if balances[from] >= amount {
		// no lock!
		balances[from] -= amount
		balances[to] += amount
	}
}


var usedCoupons = map[string]bool{}

func applyCoupon(w http.ResponseWriter, user, coupon string) {
	if !usedCoupons[coupon] {
		// Vulnerability: Does not record usage!
		balances[user] += 50
	}
}


func fetchURL(w http.ResponseWriter, r *http.Request) {
	target := r.URL.Query().Get("url")
	resp, _ := http.Get(target) // No validation: SSRF
	fmt.Fprint(w, resp.Status)
}

func main() {
	http.HandleFunc("/transfer", func(w http.ResponseWriter, r *http.Request) {
		from := r.URL.Query().Get("from")
		to := r.URL.Query().Get("to")
		amount, _ := strconv.Atoi(r.URL.Query().Get("amt"))
		transfer(from, to, amount)
	})

	http.HandleFunc("/apply-coupon", func(w http.ResponseWriter, r *http.Request) {
		user := r.URL.Query().Get("user")
		coupon := r.URL.Query().Get("coupon")
		applyCoupon(w, user, coupon)
	})

	http.HandleFunc("/ssrf", fetchURL)
}

File: python/hard.py
import requests
from threading import Thread


def get_url():
    url = input("URL to fetch: ")
    resp = requests.get(url)
    print(resp.text)


def get_invoice(user_id, invoice_id):

    invoice = db.get_invoice(invoice_id)
    print(invoice)  


class Wallet:
    def __init__(self, balance):
        self.balance = balance

    def transfer(self, target, amount):
        if self.balance >= amount:

            self.balance -= amount
            target.balance += amount


used_coupons = set()
def apply_coupon(user, coupon):
    if coupon not in used_coupons:
        user.balance -= 10



import random
def generate_session_id():
    return str(random.randint(1000, 9999))  
